<a href="https://colab.research.google.com/github/Aditya-Singla/Banknote-Authentication/blob/master/Banknote_authentication.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Importing the libraries**

In [0]:
import pandas as pd
import numpy as np

**Uploading the dataset**

In [0]:
dataset = pd.read_csv('Bank note authentication.csv')
X = dataset.iloc[:,:-1].values
y = dataset.iloc[:,-1].values



**No missing values** (*as specified by the source* https://archive.ics.uci.edu/ml/datasets/banknote+authentication )

**Splitting the dataset into training set and test set**

In [0]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size =0.2, random_state =0 )

*Checking* training and test sets

In [4]:
print(x_train)

[[-3.0866  -6.6362  10.5405  -0.89182]
 [-0.40857  3.0977  -2.9607  -2.6892 ]
 [-1.5228  -6.4789   5.7568   0.87325]
 ...
 [ 0.61652  3.8944  -4.7275  -4.3948 ]
 [ 4.0446  11.1741  -4.3582  -4.7401 ]
 [ 4.0715   7.6398  -2.0824  -1.1698 ]]


In [5]:
print(y_train)

[1 1 1 ... 1 0 0]


In [6]:
print(x_test)

[[ 4.05450e-03  6.29050e-01 -6.41210e-01  7.58170e-01]
 [ 5.13210e+00 -3.10480e-02  3.26160e-01  1.11510e+00]
 [-3.50600e+00 -1.25667e+01  1.51606e+01 -7.52160e-01]
 ...
 [-3.48100e-01 -3.86960e-01 -4.78410e-01  6.26270e-01]
 [-2.96720e+00 -1.32869e+01  1.34727e+01 -2.62710e+00]
 [-4.36670e+00  6.06920e+00  5.72080e-01 -5.46680e+00]]


In [7]:
print(y_test)

[1 0 1 0 0 0 0 0 1 1 0 0 1 0 0 0 1 1 0 0 1 0 0 1 0 0 0 1 0 0 1 0 1 0 0 0 0
 1 0 0 1 1 0 0 1 1 0 0 1 0 1 1 0 1 1 0 1 0 0 1 1 0 0 0 0 1 1 1 1 0 1 0 1 1
 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 1 1 1 1 1 0 0 1 0 0 1 0 0 0 1 0 1 1 1 0
 1 1 0 1 0 0 0 1 1 0 0 0 1 1 1 1 0 1 0 0 0 1 0 0 1 0 0 0 0 1 0 0 1 0 0 1 0
 0 1 0 0 1 0 1 0 0 0 1 0 0 1 0 1 0 1 1 0 1 1 0 1 1 1 1 0 0 0 1 1 0 0 0 0 0
 1 0 1 1 0 0 0 1 0 1 0 0 0 1 1 0 0 0 0 1 0 0 0 1 0 1 1 1 0 0 1 1 0 0 1 0 0
 0 0 0 1 1 0 0 0 0 0 0 0 1 1 0 0 1 1 0 1 1 0 1 0 1 0 0 0 0 0 1 1 1 0 0 1 1
 1 0 0 0 1 0 0 1 1 0 1 0 0 1 1 1]


In [8]:
print(y_test)

[1 0 1 0 0 0 0 0 1 1 0 0 1 0 0 0 1 1 0 0 1 0 0 1 0 0 0 1 0 0 1 0 1 0 0 0 0
 1 0 0 1 1 0 0 1 1 0 0 1 0 1 1 0 1 1 0 1 0 0 1 1 0 0 0 0 1 1 1 1 0 1 0 1 1
 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 1 1 1 1 1 0 0 1 0 0 1 0 0 0 1 0 1 1 1 0
 1 1 0 1 0 0 0 1 1 0 0 0 1 1 1 1 0 1 0 0 0 1 0 0 1 0 0 0 0 1 0 0 1 0 0 1 0
 0 1 0 0 1 0 1 0 0 0 1 0 0 1 0 1 0 1 1 0 1 1 0 1 1 1 1 0 0 0 1 1 0 0 0 0 0
 1 0 1 1 0 0 0 1 0 1 0 0 0 1 1 0 0 0 0 1 0 0 0 1 0 1 1 1 0 0 1 1 0 0 1 0 0
 0 0 0 1 1 0 0 0 0 0 0 0 1 1 0 0 1 1 0 1 1 0 1 0 1 0 0 0 0 0 1 1 1 0 0 1 1
 1 0 0 0 1 0 0 1 1 0 1 0 0 1 1 1]


**Feature Scaling** (**NOT** required for the *independent variable*)

In [0]:
from sklearn.preprocessing import StandardScaler
x_sc = StandardScaler()

x_train_sc = x_sc.fit_transform(x_train)
x_test_sc = x_sc.fit_transform(x_test)

*Checking* feature scaling

In [10]:
print(x_train_sc)

[[-1.23838989 -1.43420372  2.11416802  0.13992458]
 [-0.28386987  0.2164033  -1.02333768 -0.71665709]
 [-0.68101062 -1.40752988  1.0024975   0.9811082 ]
 ...
 [ 0.08149905  0.35150214 -1.43391934 -1.52949895]
 [ 1.30335653  1.58594301 -1.34809875 -1.69405939]
 [ 1.3129444   0.98662101 -0.81923202  0.00744699]]


**Logistic Regression**

In [0]:
from sklearn.linear_model import LogisticRegression
classifier_lr = LogisticRegression(random_state=0)
classifier_lr.fit(x_train_sc,y_train)

y_predict_lr = classifier_lr.predict(x_test_sc)


**Evaluating** Logistic Regeression 

In [12]:
from sklearn.metrics import confusion_matrix, accuracy_score, roc_auc_score
cm_lr = confusion_matrix(y_test,y_predict_lr)
print(cm_lr)
accuracy_score(y_test,y_predict_lr)


[[154   6]
 [  0 115]]


0.9781818181818182

In [13]:
auc_score_lr = roc_auc_score(y_test,y_predict_lr)
print(auc_score_lr)

0.9812500000000001


**Evaluating** K-Cross Validation Score

In [36]:
from sklearn.model_selection import cross_val_score
accuracy_lr = cross_val_score(classifier_lr, x_train_sc, y_train, cv=10 )
print( 'Accuracy:{:.2f}%'.format(accuracy_lr.mean()*100))
print( 'Standard Deviation: {:.2f}%'.format(accuracy_lr.std()*100))

Accuracy:98.27%
Standard Deviation: 0.76%


**K Nearest Neighbors**


In [0]:
from sklearn.neighbors import KNeighborsClassifier
classifier_knn = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski',p=2)
classifier_knn.fit(x_train_sc,y_train)

y_predict_knn = classifier_knn.predict(x_test_sc)

**Evaluating** K Nearest Neighbors

In [15]:
cm_knn = confusion_matrix(y_test, y_predict_knn)
print(cm_knn)
accuracy_score(y_test,y_predict_knn)

[[159   1]
 [  0 115]]


0.9963636363636363

In [16]:
auc_score_knn = roc_auc_score(y_test,y_predict_knn)
print(auc_score_knn)

0.9968750000000001


**Evaluating** K-Cross Validation Score

In [35]:
from sklearn.model_selection import cross_val_score
accuracy_knn = cross_val_score(classifier_knn, x_train_sc, y_train, cv=10 )
print( 'Accuracy:{:.2f}%'.format(accuracy_knn.mean()*100))
print( 'Standard Deviation: {:.2f}%'.format(accuracy_knn.std()*100))

Accuracy:99.82%
Standard Deviation: 0.37%


**Support Vector Machines (Kernel SVM)**

In [0]:
from sklearn.svm import SVC
classifier_svm = SVC(kernel ='rbf', random_state = 20)
classifier_svm.fit(x_train_sc,y_train)

y_predict_svm = classifier_svm.predict(x_test_sc)


**Evaluating** Kernel SVM

In [47]:
cm_svm = confusion_matrix(y_test,y_predict_svm)
print(cm_svm)
accuracy_score(y_test,y_predict_svm)

[[158   2]
 [  0 115]]


0.9927272727272727

In [48]:
auc_score_svm = roc_auc_score(y_test,y_predict_svm)
print(auc_score_svm)

0.99375


**Evaluating** K-Cross Validation Score

In [49]:
accuracy_svm = cross_val_score(classifier_svm, x_train_sc, y_train, cv=10 )
print( 'Accuracy:{:.2f}%'.format(accuracy_svm.mean()*100))
print( 'Standard Deviation: {:.2f}%'.format(accuracy_svm.std()*100))

Accuracy:100.00%
Standard Deviation: 0.00%


**Naive-Bayes Classification**

In [0]:
from sklearn.naive_bayes import GaussianNB
classifier_nb = GaussianNB()
classifier_nb.fit(x_train_sc, y_train)

y_predict_nb = classifier_nb.predict(x_test_sc)

**Evaluating** Naive Bayes

In [21]:
cm_nb = confusion_matrix(y_test, y_predict_nb)
print (cm_nb)
accuracy_score(y_test, y_predict_nb)

[[145  15]
 [ 20  95]]


0.8727272727272727

In [22]:
auc_score_nb = roc_auc_score(y_test,y_predict_nb)
print(auc_score_nb)

0.8661684782608696


**Evaluating** K-Cross Validation Score

In [40]:
accuracy_nb = cross_val_score(classifier_nb, x_train_sc, y_train, cv=10 )
print( 'Accuracy:{:.2f}%'.format(accuracy_nb.mean()*100))
print( 'Standard Deviation: {:.2f}%'.format(accuracy_nb.std()*100))

Accuracy:83.21%
Standard Deviation: 4.14%


**Decision Tree Classification**

In [0]:
from sklearn.tree import DecisionTreeClassifier
classifier_dt = DecisionTreeClassifier(criterion='entropy',random_state=0)
classifier_dt.fit(x_train_sc, y_train)

y_predict_dt = classifier_dt.predict(x_test_sc)

**Evaluating** Decision Tree Classifier

In [24]:
cm_dt = confusion_matrix(y_test, y_predict_dt)
print (cm_dt)
accuracy_score(y_test, y_predict_dt)

[[154   6]
 [  1 114]]


0.9745454545454545

In [25]:
auc_score_dt = roc_auc_score(y_test,y_predict_dt)
print(auc_score_dt)

0.9769021739130436


**Evaluating** K-Cross Validation Score

In [41]:
accuracy_dt = cross_val_score(classifier_dt, x_train_sc, y_train, cv=10 )
print( 'Accuracy:{:.2f}%'.format(accuracy_dt.mean()*100))
print( 'Standard Deviation: {:.2f}%'.format(accuracy_dt.std()*100))

Accuracy:98.54%
Standard Deviation: 0.84%


**Random Forest Classification**

In [0]:
from sklearn.ensemble import RandomForestClassifier
classifier_rf = RandomForestClassifier(n_estimators=100, criterion='entropy',random_state=0)
classifier_rf.fit(x_train_sc, y_train)

y_predict_rf = classifier_rf.predict(x_test_sc)

**Evaluating** Random Forest Classifier

In [27]:
cm_rf = confusion_matrix(y_test, y_predict_rf)
print (cm_rf)
accuracy_score(y_test, y_predict_rf)

[[157   3]
 [  0 115]]


0.9890909090909091

In [28]:
auc_score_rf = roc_auc_score(y_test,y_predict_rf)
print(auc_score_rf)

0.990625


**Evaluating** K-Cross Validation Score

In [42]:
accuracy_rf = cross_val_score(classifier_rf, x_train_sc, y_train, cv=10 )
print( 'Accuracy:{:.2f}%'.format(accuracy_rf.mean()*100))
print( 'Standard Deviation: {:.2f}%'.format(accuracy_rf.std()*100))

Accuracy:99.18%
Standard Deviation: 0.64%


**Neural Network Classifier**

P.S. This is just for fun!

In [0]:
from sklearn.neural_network import MLPClassifier
classifier_neural_network = MLPClassifier(random_state=0)
classifier_neural_network.fit(x_train_sc,y_train)

y_predict_neural_network = classifier_neural_network.predict(x_test_sc)

**Evaluating** Neural Network

In [30]:
cm_neural_network = confusion_matrix(y_test, y_predict_neural_network) 
print(cm_neural_network)
accuracy_score(y_test, y_predict_neural_network)

[[158   2]
 [  0 115]]


0.9927272727272727

In [31]:
auc_score_neural_network = roc_auc_score(y_test,y_predict_neural_network)
print(auc_score_neural_network)

0.99375


**Evaluating** K-Cross Validation Score

In [43]:
accuracy_neural_network = cross_val_score(classifier_neural_network, x_train_sc, y_train, cv=10 )
print( 'Accuracy:{:.2f}%'.format(accuracy_neural_network.mean()*100))
print( 'Standard Deviation: {:.2f}%'.format(accuracy_neural_network.std()*100))

Accuracy:100.00%
Standard Deviation: 0.00%



*Neural Network* model as well as *Kernel SVM* gave us the best overall accuracy of 99.27% !!

Well, no tuning is necessary as the accuracy has already pretty much reached the maximum.